builtin/assign_osh.py

```diff
@@ -267,31 +267,50 @@ def Run(self, cmd_val):
         return 0
 
 
-def _ReconcileTypes(rval, flag_a, flag_A, blame_word):
-    # type: (Optional[value_t], bool, bool, word_t) -> value_t
+def _ReconcileTypes(rval, flag_a, flag_A, pair, mem):
+    # type: (Optional[value_t], bool, bool, AssignArg, Mem) -> value_t
     """Check that -a and -A flags are consistent with RHS.
 
+    If RHS is empty and the current value of LHS has a different type from the
+    one expected by the -a and -A flags, we create an empty array.
+
     Special case: () is allowed to mean empty indexed array or empty assoc array
     if the context is clear.
 
     Shared between NewVar and Readonly.
+
     """
-    if flag_a and rval is not None and rval.tag() not in (
-            value_e.InternalStringArray, value_e.BashArray):
-        e_usage("Got -a but RHS isn't an array", loc.Word(blame_word))
-
-    if flag_A and rval:
-        # Special case: declare -A A=() is OK.  The () is changed to mean an empty
-        # associative array.
-        if rval.tag() == value_e.BashArray:
-            sparse_val = cast(value.BashArray, rval)
-            if bash_impl.BashArray_IsEmpty(sparse_val):
-                return bash_impl.BashAssoc_New()
-                #return bash_impl.BashArray_New()
-
-        if rval.tag() != value_e.BashAssoc:
-            e_usage("Got -A but RHS isn't an associative array",
-                    loc.Word(blame_word))
+
+    if rval is None:
+        # declare -a foo=(a b); declare -a foo; should not reset to empty array
+        if flag_a:
+            old_val = mem.GetValue(pair.var_name)
+            if old_val.tag() not in (value_e.InternalStringArray,
```

Comment:
It feels like InternalStringArray  should not be possible here -- i.e. it should not be stored in state.Mem.var_stack

But I guess that is a separate change, since it was already there

Reply:
It's possible with temporarily generatedd variables like `PIPESTATUS`. The following command hits this path with `old_val` being a temporary InternalStringArray instance which is dynamically created by `mem.GetValue`.

```console
$ bin/osh -c 'declare -a PIPESTATUS'
```